Sources/OpenAPIRuntime/Base/CopyOnWriteBox.swift

```diff
@@ -0,0 +1,190 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftOpenAPIGenerator open source project
+//
+// Copyright (c) 2023 Apple Inc. and the SwiftOpenAPIGenerator project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftOpenAPIGenerator project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+/// A reference type that wraps a value and enforces copy-on-write semantics.
+///
+/// It also enables recursive types by introducing a "box" into the cycle, which
+/// allows the owning type to have a finite size.
+@_spi(Generated)
+public final class CopyOnWriteBox<Wrapped> {
+
+    /// The stored value.
+    private var value: Wrapped
+
+    /// Creates a new box.
+    /// - Parameter value: The value to store in the box.
+    public init(value: Wrapped) {
+        self.value = value
+    }
+
+    /// Returns a copy of the stored value.
+    public func read() -> Wrapped {
+        value
+    }
+
+    /// Provides read-write access to the stored value and enforces
+    /// copy-on-write semantics.
+    /// - Parameters:
+    ///   - box: A reference to the existing box, into which a new reference
+    ///     might get written if a copy had to be made.
+    ///   - modify: The closure that modifies the value.
+    public static func write(
+        to box: inout CopyOnWriteBox<Wrapped>,
+        using modify: (inout Wrapped) -> Void
+    ) {
+        let resolvedBox: CopyOnWriteBox<Wrapped>
+        if isKnownUniquelyReferenced(&box) {
+            resolvedBox = box
+        } else {
+            resolvedBox = Self(value: box.value)
+            box = resolvedBox
+        }
+        modify(&resolvedBox.value)
+    }
```

Comment:
Why is this a static method instead of an instance method?

Reply:
Have you experimented with exposing the underlying value as computed property and enforcing uniqueness in a `_modify` block? I think this should work as well and would provide a nice interface.

Reply:
So the reason for the static func is because we need to write back into the box, and exclusivity rules prohibit the "obvious" solution of passing a ref to self and also capturing it in the closure.

I haven't tried `_modify` to avoid underscored APIs so far. Note that this is never used directly, only generated code calls this, so clarity is a bit more important here than ergonomics.

Reply:
Why aren't you passing `Value` `inout` instead of `Self`?

Reply:
Oh I see, you need to create a new box on copy. That makes sense.

Reply:
So you can make it a instance method it if you wrap the class in a struct:
```swift
public struct CopyOnWriteBox<Wrapped> {
    private final class Storage {
        var value: Wrapped
        
        init(value: Wrapped) {
            self.value = value
        }
    }
    
    private var storage: Storage
    
    public init(value: Wrapped) {
        self.storage = .init(value: value)
    }
    
    public func read() -> Wrapped {
        storage.value
    }
    
    public mutating func write(
        using modify: (inout Wrapped) -> Void
    ) {
        let resolvedStorage: Storage
        if isKnownUniquelyReferenced(&storage) {
            resolvedStorage = storage
        } else {
            resolvedStorage = Storage(value: storage.value)
            self.storage = resolvedStorage
        }
        modify(&resolvedStorage.value)
    }
}
```

Reply:
Nit: You also usually want to make these methods rethrows

Reply:
We do have a wrapper struct, that's generated. This code isn't used manually, it's SPI for the generated code, so we should consider it in the context of its call site. If you're interested in how it's used, check out the mirror PR on the generator for recursive type support.

Reply:
Resolving as this was completely changed.